In [1]:
import os
import pandas as pd
import torch

import importlib
import pickle
import re

base_path = os.getcwd()
data_path = os.path.join(base_path, "data")
question_path = os.path.join(data_path, "questions")
python_path = os.path.join(data_path, "python")

In [2]:
import ProblemSetReader
importlib.reload(ProblemSetReader)
problem_set_reader = ProblemSetReader.ProblemSetReader()
df = pd.read_pickle("./data/Geval_GPT.pkl")

In [12]:
df1 = pd.read_pickle("./data/euler_df.pkl")

In [13]:
df1

idx                                           question        answer  \
0     1  If we list all the natural numbers below $10$ ...  2.331680e+05   
1     2  Each new term in the Fibonacci sequence is gen...  4.613732e+06   
2     3  The prime factors of $13195$ are $5, 7, 13$ an...  6.857000e+03   
3     4  A palindromic number reads the same both ways....  9.066090e+05   
4     5  $2520$ is the smallest number that can be divi...  2.327926e+08   
5     6  The sum of the squares of the first ten natura...  2.516415e+07   
6     7  By listing the first six prime numbers: $2, 3,...  1.047430e+05   
7     9  A Pythagorean triplet is a set of three natura...  3.187500e+07   
8    10  The sum of the primes below $10$ is $2 + 3 + 5...  1.429138e+11   
9    15  Starting in the top left corner of a $2 \times...  1.378465e+11   
10   16  $2^{15} = 32768$ and the sum of its digits is ...  1.366000e+03   
11   19  \nYou are given the following information, but...  1.710000e+02   
12   20  $n!$ means $n \times (n - 1) \times \cdots \ti...  6.480000e+02   
13   24  A permutation is an ordered arrangement of obj...  2.783915e+09   
14   28  Starting with the number $1$ and moving to the...  6.691710e+08   
15   31  In the United Kingdom the currency is made up ...  7.368200e+04   
16   34  $145$ is a curious number, as $1! + 4! + 5! = ...  4.073000e+04   
17   35  The number, $197$, is called a circular prime ...  5.500000e+01   
18   36  The decimal number, $585 = 1001001001_2$ (bina...  8.721870e+05   
19   37  The number $3797$ has an interesting property....  7.483170e+05   
20   39  If $p$ is the perimeter of a right angle trian...  8.400000e+02   
21   41  We shall say that an $n$-digit number is pandi...  7.652413e+06   
22   48  The series, $1^1 + 2^2 + 3^3 + \cdots + 10^{10...  9.110847e+09   
23   50  The prime $41$, can be written as the sum of s...  9.976510e+05   
24   52  It can be seen that the number, $125874$, and ...  1.428570e+05   
25   56  A googol ($10^{100}$) is a massive number: one...  9.720000e+02   
26   60  The primes $3$, $7$, $109$, and $673$, are qui...  2.603300e+04   
27   62  The cube, $41063625$ ($345^3$), can be permute...  1.270360e+11   
28   63  The $5$-digit number, $16807=7^5$, is also a f...  4.900000e+01   
29   76  It is possible to write five as a sum in exact...  1.905693e+08   
30   77  It is possible to write ten as the sum of prim...  7.100000e+01   
31   79  A common security method used for online banki...  7.316289e+07   
32   80  It is well known that if the square root of a ...  4.088600e+04   
33   85  By counting carefully it can be seen that a re...  2.772000e+03   
34   94  It is easily proved that no equilateral triang...  5.184083e+08   
35  114  A row measuring seven units in length has red ...  1.647564e+10   
36  117  Using a combination of grey square tiles and o...  1.008085e+14   
37  118  Using all of the digits $1$ through $9$ and co...  4.468000e+04   
38  119  The number $512$ is interesting because it is ...  2.481558e+14   
39  132  A number consisting entirely of ones is called...  8.432960e+05   
40  142  Find the smallest $x + y + z$ with integers $x...  1.006193e+06   
41  146  The smallest positive integer $n$ for which th...  6.763333e+08   
42  160  For any $N$, let $f(N)$ be the last five digit...  1.657600e+04   
43  164  How many $20$ digit numbers $n$ (without any l...  3.781588e+14   
44  172  How many $18$-digit numbers $n$ (without leadi...  2.274853e+17   
45  173  We shall define a square lamina to be a square...  1.572729e+06   
46  178  Consider the number $45656$. \nIt can be seen ...  1.264618e+11   
47  179  Find the number of integers $1 \lt n \lt 10^7$...  9.862620e+05   
48  187  A composite is a number containing at least tw...  1.742726e+07   
49  206  Find the unique positive integer whose square ...  1.389019e+09   
50  208  A robot moves in a series of one-fifth circula...  3.319514e+17   
51  211  For a positive integer $n$, let $\sigma_2(n

In [28]:
df1['question'][17]

'The number, $197$, is called a circular prime because all rotations of the digits: $197$, $971$, and $719$, are themselves prime.\nThere are thirteen such primes below $100$: $2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79$, and $97$.\nHow many circular primes are there below one million?\n\n'

In [24]:
df['prompt'][0]

"You will be given one code answer for a math question.\n\nYour task is to evaluate the quality of this answer from three perspectives:\n\nPlease make sure you read and understand these instructions carefully. Please keep this document open while reviewing, and refer to it as needed.\n\nEvaluation Criteria:\n\nCorrectness (1-5): Does the code correctly solve the problem as described?\nEfficiency (1-5): How efficient is the code in terms of time and space complexity?\nReadability (1-5): Is the code easy to read and understand? Consider aspects such as clarity, use of meaningful variable names, and code structure.\n\nAfter run the code the result is : correct answer\n\nFor the question:\n\n\\n\\nIf we list all the natural numbers below $10$ that are multiples of $3$ or $5$, we get $3, 5, 6$ and $9$. The sum of these multiples is $23$.\nFind the sum of all the multiples of $3$ or $5$ below $1000$.\n\\n\\n.\n\nHere is the code answer for this question:\n\n\\n\\n# Python code to calculate t

In [23]:
from sympy import isprime

def rotate_number(n):
    # Rotate the digits of the number to the left by one position
    return int(str(n)[1:] + str(n)[0])

def is_circular_prime(n):
    # Get the number of digits in n
    digits = len(str(n))
    
    # Check all rotations of the number
    for _ in range(digits):
        if not isprime(n):
            return False
        n = rotate_number(n)
    
    # If all rotations are prime, return True
    return True

# List all circular primes below one million
circular_primes = [i for i in range(2, 1000000) if is_circular_prime(i)]

# Print the total number of circular primes
print(len(circular_primes))

    

135


"You will be given one code answer for a math question.\n\nYour task is to evaluate the quality of this answer from three perspectives:\n\nPlease make sure you read and understand these instructions carefully. Please keep this document open while reviewing, and refer to it as needed.\n\nEvaluation Criteria:\n\nCorrectness (1-5): Does the code correctly solve the problem as described?\nEfficiency (1-5): How efficient is the code in terms of time and space complexity?\nReadability (1-5): Is the code easy to read and understand? Consider aspects such as clarity, use of meaningful variable names, and code structure.\n\nAfter run the code the result is : correct answer\n\nFor the question:\n\n\\n\\nIf we list all the natural numbers below $10$ that are multiples of $3$ or $5$, we get $3, 5, 6$ and $9$. The sum of these multiples is $23$.\nFind the sum of all the multiples of $3$ or $5$ below $1000$.\n\\n\\n.\n\nHere is the code answer for this question:\n\n\\n\\n# Python code to calculate the sum of multiples of 3 or 5 below 1000\nsum_of_multiples = sum(x for x in range(1000) if x % 3 == 0 or x % 5 == 0)\nprint(sum_of_multiples).\n\nEvaluation Steps:\n\nCorrectness: Check if the code produces the correct output for the given problem.\nEfficiency: Analyze the code's efficiency in terms of time and space complexity.\nReadability: Evaluate how easy the code is to read and understand.\nProvide a score for each criterion from 1 to 5, where 1 is the lowest and 5 is the highest. Please only output the content after output.\n\nExample:\nCorrectness:\nEfficiency:\nReadability:\n"


In [ ]:
df1

In [20]:
df

question_id                                             prompt  \
0             1  You will be given one code answer for a math q...   
1             2  You will be given one code answer for a math q...   
2             3  You will be given one code answer for a math q...   
3             4  You will be given one code answer for a math q...   
4             5  You will be given one code answer for a math q...   
5             6  You will be given one code answer for a math q...   
6             7  You will be given one code answer for a math q...   
7             9  You will be given one code answer for a math q...   
8            10  You will be given one code answer for a math q...   
9            15  You will be given one code answer for a math q...   
10           16  You will be given one code answer for a math q...   
11           19  You will be given one code answer for a math q...   
12           20  You will be given one code answer for a math q...   
13           24  You will be given one code answer for a math q...   
14           28  You will be given one code answer for a math q...   
15           31  You will be given one code answer for a math q...   
16           34  You will be given one code answer for a math q...   
17           35  You will be given one code answer for a math q...   
18           36  You will be given one code answer for a math q...   
19           37  You will be given one code answer for a math q...   
20           39  You will be given one code answer for a math q...   
21           41  You will be given one code answer for a math q...   
22           48  You will be given one code answer for a math q...   
23           50  You will be given one code answer for a math q...   
24           52  You will be given one code answer for a math q...   
25           56  You will be given one code answer for a math q...   
26           60  You will be given one code answer for a math q...   
27           62  You will be given one code answer for a math q...   
28           63  You will be given one code answer for a math q...   
29           76  You will be given one code answer for a math q...   
30           77  You will be given one code answer for a math q...   
31           79  You will be given one code answer for a math q...   
32           80  You will be given one code answer for a math q...   
33           85  You will be given one code answer for a math q...   
34           94  You will be given one code answer for a math q...   
35          114  You will be given one code answer for a math q...   
36          117  You will be given one code answer for a math q...   
37          118  You will be given one code answer for a math q...   
38          119  You will be given one code answer for a math q...   
39          132  You will be given one code answer for a math q...   
40          142  You will be given one code answer for a math q...   
41          146  You will be given one code answer for a math q...   
42          160  You will be given one code answer for a math q...   
43          164  You will be given one code answer for a math q...   
44          172  You will be given one code answer for a math q...   
45          173  You will be given one code answer for a math q...   
46          178  You will be given one code answer for a math q...   
47          179  You will be given one code answer for a math q...   
48          187  You will be given one code answer for a math q...   
49          206  You will be given one code answer for a math q...   
50          208  You will be given one code answer for a math q...   
51          211  You will be given one code answer for a math q...   
52          222  You will be given one code answer for a math q...   
53          231  You will be given one code answer for a math q...   
54          249  You will be given one code answer for a math q...   
55          250  You will be given one code answer for a math q...   
56          35

# Test GPT

In [4]:
import GPTAgent
importlib.reload(GPTAgent)
gpt_agent = GPTAgent.GPTAgent()

ModuleNotFoundError: No module named 'g4f'

In [126]:
response_string = gpt_agent.generate_chain_of_thoughts(problem_set_reader.get_question(125))
procedures = response_string.split("\n")
procedures

['1. Define the problem statement and understand the required input/output.',
 '2. Generate palindromic numbers from 1 to $10^8$.',
 '3. For each palindromic number, find all possible combinations of consecutive squares that sum up to the palindromic number.',
 '4. Verify if the sum of consecutive squares is valid for each palindromic number, excluding the case of $1 = 0^2 + 1^2$.',
 '5. Print the sum of all numbers less than $10^8$ that are both palindromic and can be written as the sum of consecutive squares.']

In [130]:
from g4f.client import Client

client = Client()
response = client.chat.completions.create(
    model="gpt-4",
    messages=[{"role": "system", "content": "Only respond in English."},
              {"role": "user", "content": "How to make a cup of tea?"}, 
              {"role": "assistant", "content": "Simply boil water and add tea."}, 
              {"role": "user", "content": "What is the last question I asked you?"}],
)
response.choices[0].message.content

"I'm sorry, but I don't have the ability to recall past interactions or questions from previous conversations. Each session with me is independent, and I treat every interaction as a new encounter. How may I assist you today?\n"

In [127]:
messages = []

for step_description in procedures:
    prompt = f"Generate at most two line of python code (packed by ```python {{python_code}}```) for this step (and only this step): {step_description}. \
        Directly output the code without explaination, output space to indent if necessary"

    messages.append({"role": "user", "content": prompt})
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=messages
    )
    ret_string = response.choices[0].message.content
    print(ret_string)
    messages.append({"role": "assistant", "content": ret_string})

```
python
problem_statement = input("请输入问题描述：")
required_input_output = input("请输入所需输入输出：")
```
```python
for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1]:
        print(i)
```
```python
import itertools
for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1]:
        for r in range(1, int(i**0.5) + 1):
            for combo in itertools.combinations(range(1, r + 1), r):
                if sum(j**2 for j in combo) == i:
                    print(f"{i} = {'+'.join(str(j**2) for j in combo)}")
```
```python
for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1] and i!= 2 and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)):
        print(i)
```
```python
sum_of_palindromes = sum(i for i in range(1, 10**8 + 1) if str(i) == str(i)[::-1] and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)) and i != 1)
print(sum_of_palindromes)
```


In [128]:
raw_code = [messages['content'] for messages in messages if messages['role']=='assistant']
raw_code

['```\npython\nproblem_statement = input("请输入问题描述：")\nrequired_input_output = input("请输入所需输入输出：")\n```',
 '```python\nfor i in range(1, 10**8 + 1):\n    if str(i) == str(i)[::-1]:\n        print(i)\n```',
 '```python\nimport itertools\nfor i in range(1, 10**8 + 1):\n    if str(i) == str(i)[::-1]:\n        for r in range(1, int(i**0.5) + 1):\n            for combo in itertools.combinations(range(1, r + 1), r):\n                if sum(j**2 for j in combo) == i:\n                    print(f"{i} = {\'+\'.join(str(j**2) for j in combo)}")\n```',
 '```python\nfor i in range(1, 10**8 + 1):\n    if str(i) == str(i)[::-1] and i!= 2 and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)):\n        print(i)\n```',
 '```python\nsum_of_palindromes = sum(i for i in range(1, 10**8 + 1) if str(i) == str(i)[::-1] and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)) and i != 1)\nprint

In [129]:
pattern = r"```python(.*?)```"
code_cleansed = [match.group(1) for match in re.finditer(pattern, '\n'.join(raw_code), re.DOTALL)]
code_concat = "".join(code_cleansed)
print(code_concat)


for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1]:
        print(i)

import itertools
for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1]:
        for r in range(1, int(i**0.5) + 1):
            for combo in itertools.combinations(range(1, r + 1), r):
                if sum(j**2 for j in combo) == i:
                    print(f"{i} = {'+'.join(str(j**2) for j in combo)}")

for i in range(1, 10**8 + 1):
    if str(i) == str(i)[::-1] and i!= 2 and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)):
        print(i)

sum_of_palindromes = sum(i for i in range(1, 10**8 + 1) if str(i) == str(i)[::-1] and any(sum(j**2 for j in range(a, b)) == i for a in range(1, int(i**0.5) + 1) for b in range(a, int(i**0.5) + 2)) and i != 1)
print(sum_of_palindromes)



You will be given one code answer for a math question.

Your task is to evaluate the quality of this answer from three perspectives:

Please make sure you read and understand these instructions carefully. 

Evaluation Criteria:

Correctness： Measures whether the code solves the problem as expected. If the code provides correct results and has no logical errors, it can be given a high score. If the code has errors or gives incorrect results, a low score should be given. Score range: 0-5.

Efficiency: Measures the efficiency of the code in terms of time and space. This includes time complexity and space complexity. If the code has low time and space complexity, it can be given a high score. A low score should be given if the code has high time or space complexity. Score range: 0-5.

Readability:Measures how easy the code is to read and understand. This includes things like the clarity of variable names, the structure of the code, and the quality of comments. If the code is easy to read and understand, it will score high. If the code is difficult to understand or the structure is confusing, a low score should be given. Score range: 0-5.


For the question:
You will be given one code answer for a math question.

Your task is to evaluate the quality of this answer from three perspectives:

Please make sure you read and understand these instructions carefully. 

Evaluation Criteria:

Correctness： Measures whether the code solves the problem as expected. If the code provides correct results and has no logical errors, it can be given a high score. If the code has errors or gives incorrect results, a low score should be given. Score range: 0-5.

Efficiency: Measures the efficiency of the code in terms of time and space. This includes time complexity and space complexity. If the code has low time and space complexity, it can be given a high score. A low score should be given if the code has high time or space complexity. Score range: 0-5.

Readability:Measures how easy the code is to read and understand. This includes things like the clarity of variable names, the structure of the code, and the quality of comments. If the code is easy to read and understand, it will score high. If the code is difficult to understand or the structure is confusing, a low score should be given. Score range: 0-5.


For the question:
\nThe number, $197$, is called a circular prime because all rotations of the digits: $197$, $971$, and $719$, are themselves prime.\nThere are thirteen such primes below $100$: $2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79$, and $97$.\nHow many circular primes are there below one million?\n

Here is the code answer for this question:
\n from sympy import isprime\n\ndef rotate_number(n):\n return int(str(n)[1:] + str(n)[0])\n\ndef is_circular_prime(n):\n digits = len(str(n))\n for _ in range(digits):\n if not isprime(n):\n return False\n n = rotate_number(n)\n return True\n\ncircular_primes = [i for i in range(2, 1000000) if is_circular_prime(i)]\nprint(len(circular_primes)).\n

Evaluation Steps:
Start: Initialize Correctness, Efficiency, and Readability to 100, and you should set a scoring guideline for yourself!

Correctness: Checks that the code produces the correct output for the given problem. If you make a mistake somewhere, deduct points, and you can decide how many points to deduct based on the size of the mistake.

Efficiency: Analyze the efficiency of the code in terms of time and space complexity. If the time-space complexity is high, a larger number of points should be deducted.

Readability: Evaluate how easy the code is to read and understand.

Provide a score for each criterion from 1 to 5, where 1 is the lowest and 5 is the highest. Do not output detail descrivetion, Do not output detail descrivetion, Please only output the content after output.

Example:
Correctness:
Efficiency:
Readability:

Here is the code answer for this question:
\nfrom sympy import isprime\n\ndef rotate_number(n):\n return int(str(n)[1:] + str(n)[0])\n\ndef is_circular_prime(n):\n digits = len(str(n))\n for _ in range(digits):\n if not isprime(n):\n return False\n n = rotate_number(n)\n return True\n\ncircular_primes = [i for i in range(2, 1000000) if is_circular_prime(i)]\nprint(len(circular_primes)).\n

Evaluation Steps:
Start: Initialize Correctness, Efficiency, and Readability to 100, and you should set a scoring guideline for yourself!

Correctness: Checks that the code produces the correct output for the given problem. If you make a mistake somewhere, deduct points, and you can decide how many points to deduct based on the size of the mistake.

Efficiency: Analyze the efficiency of the code in terms of time and space complexity. If the time-space complexity is high, a larger number of points should be deducted.

Readability: Evaluate how easy the code is to read and understand.

Provide a score for each criterion from 1 to 5, where 1 is the lowest and 5 is the highest.  Please only output the content after output.

Example:
Correctness:
Efficiency:
Readability: